ci: Fix "macOS native" job

hebasto opened the pull request:
Homebrew [promoted](https://github.com/Homebrew/homebrew-core/pull/150390) `python@3.12` to the default `python3`. Now, our "macOS native" CI job is facing the following issues:

1. Installing `qt@5` [requires](https://github.com/bitcoin/bitcoin/actions/runs/8216848118/job/22471875454#step:4:51) re-installing `python@3.12`:
```
==> Fetching dependencies for qt@5: readline, python@3.12 and gettext
```
2. Re-installing `python@3.12` [fails](https://github.com/bitcoin/bitcoin/actions/runs/8216848118/job/22471875454#step:4:127) due to symbolic link conflicts on macOS `x86_64`:
```
==> Pouring python@3.12--3.12.2_1.ventura.bottle.tar.gz
Error: The `brew link` step did not complete successfully
```
3. Homebrew's `python@3.12` is marked as externally managed (according to PEP 668), necessitating different approaches for installing Python packages.

This pull request resolves all the issues mentioned above.

hebasto committed:
ci: Add workaround for Homebrew's python link error

Promoting Homebrew's python@3.12 to the default python3 breaks symbolic
links on macOS x86_64.

This change adds a workaround for that issue.

Also see: https://github.com/actions/runner-images/issues/9471 etc.

hebasto committed:
ci, macos: Use `--break-system-packages` with Homebrew's python

Homebrew's python@3.12 is marked as externally managed (PEP 668),
necessitating different approaches for installing Python packages.

For more details, please refer to https://github.com/orgs/Homebrew/discussions/3404.

m3dwards commented:
ACK https://github.com/bitcoin/bitcoin/pull/29610/commits/eca024fa97de1500bf2a176cfcfe3177685fab0c

Don't have a x86 Mac to hand right now so ran mac cross CI job to test.

This works but two alternative approaches to the pip install problem could have been to either use a python virtual environment (venv) or to pass `--break-system-packages` when pip installing. As this is a throwaway build runner venv is probably overkill.

fanquake commented:
It does seem not ideal to have to forcefully install one verison of Python, only to then not be able to use it for anything Python related, and then also refactor the CI to use a different (unclear which?) version of Python.

hebasto commented:
@m3dwards @maflcko @fanquake 

Thank you for your reviews!

> > `--break-system-packages`
> 
> If this causes a smaller diff, I'd say this is preferred.

Implemented.

> It does seem not ideal to have to forcefully install one verison of Python, only to then not be able to use it for anything Python related, and then also refactor the CI to use a different (unclear which?) version of Python.

Reworked, so Homebrew's python is still using.

maflcko reviewed (approved):
lgtm

m3dwards commented on ci/test/00_setup_env_mac_native.sh:
```diff
@@ -7,7 +7,9 @@
-export PIP_PACKAGES="zmq"
+# Homebrew's python@3.12 is marked as externally managed (PEP 668).
+# Therefore, `--break-system-packages` is needed.
+export PIP_PACKAGES="--break-system-packages zmq"
```
Agreed. It's probably going to start coming up in other jobs as python and base containers get upgraded.

m3dwards commented:
reACK https://github.com/bitcoin/bitcoin/pull/29610/commits/acc06bc91f80ddf4e015dcdf0b984bbdbfcb5ca3 to get the CI passing again.

Contrary to my previous comment, when making a more global change to CI perhaps the more permanent and proper thing would be to use a venv.

fanquake commented:
Backported to 27.x in #29693 and 26.x in #29719.
